src/image_host/custom_image_host.cpp

```diff
@@ -0,0 +1,225 @@
+
+    for (auto it = root_obj.begin(); it != root_obj.end(); ++it)
+    {
+        const QString distro_name = it.key();
+        const QJsonObject distro_obj = it.value().toObject();
+
+        QStringList aliases = distro_obj.value("aliases").toString().split(",", Qt::SkipEmptyParts);
+        for (QString& alias : aliases)
+            alias = alias.trimmed();
+
+        images.push_back(mp::VMImageInfo{aliases,
+                                         distro_obj["os"].toString(),
+                                         distro_obj["release"].toString(),
+                                         distro_obj["release_codename"].toString(),
+                                         distro_obj["release_title"].toString(),
+                                         true,
+                                         distro_obj["items"][arch]["image_location"].toString(),
+                                         distro_obj["items"][arch]["id"].toString(),
+                                         "",
+                                         distro_obj["items"][arch]["version"].toString(),
+                                         distro_obj["items"][arch]["size"].toInt(-1),
+                                         true});
+    }
+
+    return images;
+}
+} // namespace
+
+mp::CustomManifest::CustomManifest(std::vector<VMImageInfo>&& images)
+    : products{std::move(images)}, image_records{map_aliases_to_vm_info(products)}
+{
+}
+
+mp::CustomVMImageHost::CustomVMImageHost(URLDownloader* downloader)
+    : BaseVMImageHost{downloader},
+      arch{QSysInfo::currentCpuArchitecture()},
+      manifest{},
+      remote{no_remote}
+{
+}
+
+std::optional<mp::VMImageInfo> mp::CustomVMImageHost::info_for(const Query& query)
+{
+    auto custom_manifest = manifest_from(query.remote_name);
+
+    auto it = custom_manifest->image_records.find(query.release);
+
+    if (it == custom_manifest->image_records.end())
+        return std::nullopt;
+
+    return *it->second;
+}
+
+std::vector<std::pair<std::string, mp::VMImageInfo>> mp::CustomVMImageHost::all_info_for(
+    const Query& query)
+{
+    std::vector<std::pair<std::string, mp::VMImageInfo>> images;
+
+    auto image = info_for(query);
+    if (image != std::nullopt)
+        images.push_back(std::make_pair(query.remote_name, *image));
+
+    return images;
+}
+
+std::vector<mp::VMImageInfo> mp::CustomVMImageHost::all_images_for(const std::string& remote_name,
+                                                                   const bool allow_unsupported)
+{
+    std::vector<mp::VMImageInfo> images;
+    auto custom_manifest = manifest_from(remote_name);
+    std::copy(custom_manifest->products.begin(),
+              custom_manifest->products.end(),
+              std::back_inserter(images));
+
+    return images;
+}
+
+std::vector<std::string> mp::CustomVMImageHost::supported_remotes()
+{
+    return std::vector<std::string>{remote};
```

Comment:
We could omit the `std::vector<std::string>`. Either just `return remote;` or `return {remote};` if you want to make it more obvious that it's a copy.